testutils: fix NoData behavior

pjonsson opened the pull request:
### Reason for this pull request

Rasterio 1.4.4 and older relies on
an absent NoData value being interpreted
as NoData=0 by GDAL, which changed
in GDAL 3.11.

Update the test code to behave the same
way across all GDAL versions.

### Proposed changes

- 
- 



 - [ ] Closes #xxxx
 - [ ] Tests added / passed
 - [X] Pull Request Title will make sense in ODC Release Notes

<!--
See https://github.com/blog/2111-issue-and-pull-request-templates for more
information on pull request templates.

-->


<!-- readthedocs-preview opendatacube start -->
----
📚 Documentation preview 📚: https://opendatacube--2049.org.readthedocs.build/en/2049/

<!-- readthedocs-preview opendatacube end -->

pjonsson committed:
testutils: fix NoData behavior

Rasterio 1.4.4 and older relies on
an absent NoData value being interpreted
as NoData=0 by GDAL, which changed
in GDAL 3.11.

Update the test code to behave the same
way across all GDAL versions.

pjonsson commented:
This is arguably a Rasterio bug that went undiscovered because of a longstanding bug before GDAL 3.11.0. Since there isn't a Rasterio release on the horizon, we make sure to always set the NoData value to something so we don't rely on the current behavior of Rasterio.

This PR just changes test code, but it happens to fix 50% of the failing tests with GDAL 3.11.

SpacemanPaul reviewed (commented):
What about floating point bands? `0` is a sensible nodata default for integer bands, but floating point bands should use `NaN`.

As noted in Emma's PR, `testutils.io` is not just used in tests, so worth handling this case there.

pjonsson commented:
> What about floating point bands? `0` is a sensible nodata default for integer bands, but floating point bands should use `NaN`.

Fixed, and rebased on top of latest develop.

SpacemanPaul reviewed (approved):
Great, thanks.
